src/config.spec.js

```diff
@@ -1,43 +1,167 @@
+const path = require('path')
+
+const virtualConfig = jest.fn().mockReturnValue({})
+let config
+let fs
+
+beforeEach(() => {
+  jest.clearAllMocks()
+
+  // Reset the config/context/state singletons with jest.resetModules().
+  // Unfortunately this seems to destroy mocks from the main scope
+  // so we have to redefine them for each test; I tried using
+  // `delete require.cache[require.resolve('./config')]`
+  // but Jest overrides the NodeJS require-mechanism and
+  // refuses to let you clear individual entries from the cache.
+  jest.resetModules()
+  jest.doMock('./utils/fs')
+  fs = require('./utils/fs')
+
+  // Emulate a config file
+  const configPath = './config.virtual'
+  fs.findUpwardsFile.mockResolvedValue(configPath)
+  jest.doMock(configPath, virtualConfig, { virtual: true })
+
+  config = jest.requireActual('./config')
```

Comment:
This part could probably be done more efficiently if we move the require-ing of the dynamic config file path to a separate function from the parser, but a lot of frustration went into this solution and I'm still a bit angry with Jest about it so it'll stay for now.